### DISCUSSION TITLE
CPU Usage Spikes

### DISCUSSION THREAD

**Discussion by taavetmaask** [pocketbase/pocketbase#3714] [Q&A]
Hello everyone,

I've encountered issues with Pocketbase after upgrading from version 0.8 to 0.19. While the data migration proceeded smoothly, I am experiencing performance issues - CPU and Disk IO spikes. Typically, a server restart is required to resolve the problem.

<img width="829" alt="Screenshot 2023-11-08 at 19 57 42" src="https://github.com/pocketbase/pocketbase/assets/26470755/ce0165e3-abfc-4e07-93bd-f4fca885672a">

In version 0.8, this issue was never present. Moreover, the performance problems do not consistently correspond with network traffic. I hope someone else has encountered this and can offer guidance or a solution.

Thank you in advance!

**Comment by ganigeorgiev** [comment]
Is PocketBase the only application running on the server?

A lot has changed between v0.8 and v0.19 and the performance in the latest release should be definitely better from releases before v0.10.0.

Without a minimal reproducible repo or at least some details what actually is being executed I'm not sure how to help you debug it.

I don't think this is a PocketBase issue as I'm not able to reproduce it neither locally nor on the live demo. Here is a screenshot from the Hetzner live demo dashboard for the past week:

![hetzner_dashboard](https://i.imgur.com/1NGtV63.png)

There is definitely something else going on in your environment but I'm not sure what exactly is causing the cpu usage (it could be bulk thumbs generations, slow query, conflicting other running service, etc.).

**Reply by ganigeorgiev** [reply]
And also what do you mean by _" Typically, a server restart is required to resolve the problem."_

Does the CPU usage never get down on its own after some time?

The memory usage may remain high after bursty/spiky scenarios since the GC may not release entirely the claimed memory if there is no need for (eg. no memory pressure from the OS), but CPU and Disk IO definitely should go back to its normal idle state if the PocketBase server doesn't do anything.

**Reply by taavetmaask** [reply]
Yes, it’s the only service running on my server. It does eventually recover on its own, but the process takes over five/ten minutes. Usually, I've restarted the server by that time.

**Reply by ganigeorgiev** [reply]
There is really not enough details to help you debug it and I doubt that it is caused by the upgrade.

CPU and Disk IO intensive operations in Pocketbase are usually:
- thumbs generation (_they are generated lazily on the first access_)
- execution of aggregations or complex queries against large datasets
- heavy logs write (_this will be improved in the next release with the logs generalization; you can disable the logs from the "Admin UI > Settings > Application > Logs max days retention" input and set it to `0`_)
- user login/register (_this is mostly CPU only due to the bcrypt hash generation_)

_edit: added user login/register to the list of possible cpu/disk intensive tasks_
